### PR TITLE
enhancement proposal : customization of module files 

### PR DESCRIPTION
**Pull request by alalazo**
##### Objective 
Being able to customize what gets written in a module file

##### TLDR
Currently the modifications that gets written in a module file are:

- generic modifications to some list of paths in the environment

These modifications are obtained inspecting `pkg.prefix` according to the same set of rules for every package. I would like to extend this process allowing:

- package-specific modifications to the environment (this PR)
- site-specific modifications via a configuration file (a subsequent PR)

##### Current Status
- [x] class that manages environment modifications during both installation and module file creation (module + unit tests)
- [x] environment variables set in a package are written to tcl or dotkit module files
- [x] polished the interface that package maintainers will need to use
- [x] validation of environment modifications (i.e. detect conflicting changes to the same variable issued by different packages)

##### Examples of use
1. `openmpi` and `mpich` packages in this PR set the environment variables for compiler wrappers differently depending on whether they are used as a dependency during an installation or outside of spack via module files

##### Details
The biggest change done in this PR is the introduction of a class (`EnvironmentModifications`) to manage environment modifications both during install and module file creation. An object of this class is essentially a FIFO queue containing requests to modify the environment. Those requests can be executed at a later time and from a scope other than where they were created (in which case the environment is modified and the list is cleared.) 

Alternatively the requests can be inspected and processed. This is what is done during module file creation when each module file writer (`TclModule`, `Dotkit`) maps each single request to a formatted line and writes that line to a file.

An interesting side effect of this design is the possibility to inspect the whole set of requests done by different packages before applying them. This could permit to reveal inconsistencies (e.g.  variables that are set by a package and unset by another) and output diagnostic messages to the user.

### CODE REVIEW

**Comment by alalazo** [comment]
@tgamblin : could you please have a look at the skeleton of the thing and tell me if the overall idea fits what you had in mind?

**Comment by citibeth** [comment]
Looks likes a well-thought out idea so far.  Can you please explain more about the use case and the problem(s) it solves?  Thanks!

**Comment by alalazo** [comment]
@citibeth One of the most requested features (and one that we need at our site) is the possibility to 'customize' the modulefiles that are written after a successfull installation. The customizations in which I am interested here will consist in the addition of new entries and/or in the modification of the usual entries that are added via automatic discovery of some standard directories. Also, these customizations may happen either per-package (most part of the cases) or per site.

As an example of a per-package customization you may think of an MPICH installation that could set `MPICH_CC` or similar variables in the modulefiles. IntelMPI would instead represent an example of a per-site customization, as the installation will be tuned to the right network fabrics, etc. setting a bunch of environment variables appropriately.

To reach the goal without creating a tight coupling among different parts of `spack` a viable approach would be to (imo) :
- group together the modifications that are done to the environment (and are now plain `os.environ[name]` assignment within some function) into an object
- build a mechanism that constructs one of these objects appropriately and produces the desired customizations (in another PR?)

File: lib/spack/spack/modules.py [hunk 1]
```diff
@@ -79,8 +74,35 @@ def print_help():
...
+            for name in names:
+                variables = PathInspector.dirname2varname.get(name, None)
+                if variables is None:
+                    continue
+                absolute_path = join_path(os.path.abspath(directory), name)
+                for variable in variables:
+                    env.prepend_path(variable, absolute_path)
+
+    env, inspector = EnvironmentModifications(), PathInspector()
+    os.path.walk(path, inspector, env)
```
Comment (tgamblin) [inline lib/spack/spack/modules.py]: Do you really want to walk the whole prefix?  For most of these I think just checking at the top level would be fine.  For pkgconfig, there are a few standard locations you could check (`lib/pkgconfig`, `lib64/pkgconfig`) but I'm not sure you wan to grab all of them unconditionally.  I think for some packages this might pick up stuff you don't want, like tests and such.  I would leave adding the deeper directories to the package implementor --- they can extend it if they want.

File: lib/spack/spack/modules.py [hunk 2]
```diff
@@ -88,67 +110,40 @@ def __init__(cls, name, bases, dict):
...
+
+        # long description is the docstring with reduced whitespace.
+        self.long_description = None
+        if self.spec.package.__doc__:
+            self.long_description = re.sub(r'\s+', ' ', self.spec.package.__doc__)
+
+    # @property
+    # def paths(self):
+    #         # Add python path unless it's an actual python installation
+    #         # TODO : is there a better way to do this?
```
Comment (tgamblin) [inline lib/spack/spack/modules.py]: I think a better way might be embedding some of this in the Python package itself -- you could have the Python package add a custom `PYTHONPATH` for Python and/or Python extensions.  Might be good to have a package callback for that -- Ruby could do the same.

File: lib/spack/spack/modules.py [hunk 3]
```diff
@@ -88,67 +110,40 @@ def __init__(cls, name, bases, dict):
             if cls.name != 'env_module':
                 module_types[cls.name] = cls
 
-
     def __init__(self, spec=None):
         # category in the modules system
         # TODO: come up with smarter category names.
```
Comment (tgamblin) [inline lib/spack/spack/modules.py]: :+1:

File: lib/spack/spack/modules.py [hunk 4]
```diff
@@ -156,18 +151,27 @@ def write(self):
...
 
-
-    def _write(self, stream):
-        """To be implemented by subclasses."""
+    def write_header(self, stream):
         raise NotImplementedError()
 
+    def process_environment_command(self, env):
+        for command in env:
+            # FIXME : how should we handle errors here?
```
Comment (tgamblin) [inline lib/spack/spack/modules.py]: If you really want to get fancy, you could call `inspect.currentframe()` from within `prepend_path`, etc., and store the file/line info from the place where the modification was created.  That would let you tell the user where they screwed up.

File: lib/spack/spack/package.py [hunk 5]
```diff
@@ -983,28 +977,42 @@ def module(self):
                           fromlist=[self.__class__.__name__])
 
 
-    def setup_dependent_environment(self, module, spec, dependent_spec):
-        """Called before the install() method of dependents.
+    def environment_modifications(self, dependent_spec):
```
Comment (tgamblin) [inline lib/spack/spack/package.py]: I think this should just take an EnvironmentModifications() as a parameter, and not require the package author to know how to create one.  e.g.:

```python
def setup_environment(self, env):
    env.prepend_path(...)
    env.append_path(...)
```

Also, I think the dependent_spec aspect is kind of confusing.  It may be that we need two functions:
```python
def setup_environment(self, env):
    """hook in a package to set up *its own* environment for anything that wants to use it."""
    # most normal module stuff goes here

def setup_dependent_environment(self, env, dependent_spec):
    """hook in a package that allows it to set up the environment specifically for *dependent* packages."""
    # stuff that python might do, like add a site-packages directory to the `PYTHONPATH` on behalf
    # of an extension, should go here.
```

I think this separates the things that need to go into a normal module from the things that need to go into the *dependent's* module.  When you generate the module file for a dependent, you would probably want to include env modifications from its *dependencies'* `setup_dependent_environment` functions and from *its* `setup_environment` function.

Does that seem reasonable?

File: lib/spack/spack/package.py [hunk 6]
```diff
@@ -983,28 +977,42 @@ def module(self):
...
-           point to the lib/pythonX.Y/site-packages directory in the
-           module's install prefix.  This could set that variable.
+        2. A lot of Qt extensions need `QTDIR` set.  This can be used to do that.
 
-        2. Extensions often need to invoke the 'python' interpreter
-           from the Python installation being extended.  This routine can
-           put a 'python' Execuable object in the module scope for the
-           extension package to simplify extension installs.
+        Args:
+            dependent_spec: dependent (or extension) of this spec
```
Comment (tgamblin) [inline lib/spack/spack/package.py]: I'm wondering whether we should make `setup_dependent_environment` into `setup_extension_environment`.  It is really only for extensions that a dependency (python) knows how to make modifications to its extensions' environment, anyway.  I am not sure it actually makes sense generally when you have a generic `setup_environment` method as I suggested above.

File: lib/spack/spack/package.py [hunk 7]
```diff
@@ -983,28 +977,42 @@ def module(self):
...
+
+        Returns:
+            instance of environment modifications
+        """
+        return EnvironmentModifications()
+
+    def module_modifications(self, module, spec, dependent_spec):
```
Comment (tgamblin) [inline lib/spack/spack/package.py]: I like separating out the module modifications, though these could easily be combined with `setup_extension_environment`

File: lib/spack/spack/util/environment.py [hunk 8]
```diff
@@ -40,11 +40,13 @@ def env_flag(name):
     return False
 
 
+# FIXME : remove this function ?
```
Comment (tgamblin) [inline lib/spack/spack/util/environment.py]: I think it makes sense to keep these, as there are still good reasons a package developer might want to use them in `install()`.  Mainly, if they want to make environment changes *only* for the build that *won't* go into a module. Modifications to `CC`, `CXX`, etc. probably fall into this category most of the time.

File: lib/spack/spack/package.py (hunk 5 above)
Comment (alalazo) [inline lib/spack/spack/package.py]: @tgamblin Very reasonable. We may in fact have a default implementation for `setup_dependent_environment` that delegates to `setup_environment`, so that in all the cases where the modifications are exactly the same, the user can implement just one of the two functions.

Regarding the parameter vs. return value : I usually prefer the second way of doing things, but I see the point here. I'll make these changes asap.

File: var/spack/repos/builtin/packages/openmpi/package.py [hunk 9]
```diff
@@ -41,12 +41,14 @@ class Openmpi(Package):
...
 
-    def setup_dependent_environment(self, module, spec, dep_spec):
-        """For dependencies, make mpicc's use spack wrapper."""
-        os.environ['OMPI_CC'] = 'cc'
-        os.environ['OMPI_CXX'] = 'c++'
-        os.environ['OMPI_FC'] = 'f90'
-        os.environ['OMPI_F77'] = 'f77'
+    def environment_modifications(self, dependent_spec):
+        env = super(Openmpi, self).environment_modifications(dependent_spec)
+        # FIXME : the compilers should point to the current wrappers, not to generic cc etc.
```
Comment (tgamblin) [inline var/spack/repos/builtin/packages/openmpi/package.py]: This gets complicated.  These variables are intended to be set to *Spack's* wrappers during a Spack build, so that dependents using MPI implementations properly RPATH their dependencies.  I have recently put in changes that let you refer to these more easily, so you could do this:

```python
        env.set_env('OMPI_CC', spack_cc)
        env.set_env('OMPI_CXX', spack_cxx)
        env.set_env('OMPI_FC', spack_fc)
        env.set_env('OMPI_F77', spack_f77)
```

However, if they are loaded by a module *outside of spack*, they should probably point to the real compilers.  So you want to point at these:

```python
        env.set_env('OMPI_CC', self.compiler.cc)
        env.set_env('OMPI_CXX', self.compiler.cxx)
        env.set_env('OMPI_FC', self.compiler.fc)
        env.set_env('OMPI_F77', self.compiler.f77)
```

This could be made more consistent/intuitive (sorry!) but that's what you want at the moment.  Maybe we could add `real_cc`, `real_cxx`, or similar later, but let's ignore that for now.

I'm not sure how best to do this -- the EnvModifications *may* need a way to indicate that different things should be done inside Spack and outside.  Or maybe there is a cleaner way to do it... I *think* this is only an issue for things that wrap compilers, or at least that is where I have encountered it so far.  Suggestions welcome!

File: lib/spack/spack/package.py (hunk 6 above)
Comment (alalazo) [inline lib/spack/spack/package.py]: @tgamblin It could make sense for things like MPI libraries, that have to set `MPI_CC` like variables to the compiler wrapper when installing and *may* want to set them to the real compiler in modules

File: var/spack/repos/builtin/packages/python/package.py [hunk 10]
```diff
@@ -89,8 +89,19 @@ def python_include_dir(self):
...
+    def environment_modifications(self, extension_spec):
+        env = super(Python, self).environment_modifications(extension_spec)
+        if extension_spec is not None:
+            # Set PYTHONPATH to include site-packages dir for the
+            # extension and any other python extensions it depends on.
+            python_paths = []
+            for d in extension_spec.traverse():
+                if d.package.extends(self.spec):
+                    python_paths.append(os.path.join(d.prefix, self.site_packages_dir))
+            env.set_env['PYTHONPATH'] = ':'.join(python_paths)
```
Comment (tgamblin) [inline var/spack/repos/builtin/packages/python/package.py]: Tricky part here is that this is actually setting up the `dependent_spec`'s environment.  So how do you know what goes in `python`'s module and what goes in, e.g., `numpy`'s module?  I think you don't, so we should separate this out like I suggested above.

File: lib/spack/spack/modules.py (hunk 1 above)
Comment (alalazo) [inline lib/spack/spack/modules.py]: @tgamblin Ok, will restore the previous logic. It seems in fact better to leave the responsibility of adding non-standard directories to the package maintainer.

File: lib/spack/spack/environment.py [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+import os.path
+import collections
+
+
+class AttributeHolder(object):
+    """
+    Policy that permits to store any kind of attribute on self. The attributes must be passed as key/value pairs
+    during the initialization of the instance.
+    """
+    def __init__(self, **kwargs):
```
Comment (tgamblin) [inline lib/spack/spack/environment.py]: Are the `kwargs` here ever used?  I don't think so -- maybe this is too generic?

I like having a class per type-of-modification, but I think you want something more like this:

```python
class NameModifier(object):
    def __init__(self, name):
        self.name = name

class NameValueModifier(object):
    def __init__(self, name, value):
        self.name = name
        self.value = value

class SetEnv(NameValueModifier):
    def execute(self):
        os.environ[self.name] = str(self.value)

class UnsetEnv(NameModifier):
    def execute(self):
        os.environ.pop(self.name, None)
```

Now the subclasses are much shorter and more concise...

File: lib/spack/spack/modules.py [hunk 12]
```diff
@@ -156,18 +151,27 @@ def write(self):
...
+            yield self.formats[type(command)].format(command)
```
Comment (tgamblin) [inline lib/spack/spack/modules.py]: Per my idea for the `AttributeHolder` above, you could make this read:

```python
    yield self.formats[type(command)].format(**command.format_args)
```

so that formats could be prettier:

    formats = {
        PrependPath: 'dk_alter {name} {path}\n',
        SetEnv: 'dk_setenv {name} {value}\n'
    }

File: lib/spack/spack/environment.py (hunk 11 above)
Comment (alalazo) [inline lib/spack/spack/environment.py]: @tgamblin I thought of the `kwargs` as a way to attach attributes to commands. The idea was to make it possible to write something like:
```
env.set_env('A', value, who=spec, where=...)
```
in client code and being able to refer to the additional information from functions written by the same client.

**Comment by tgamblin** [comment]
@alalazo: ok I went over this.  It looks really good!  I like the design; I think this is definitely the right direction.  High level, my comments are basically:

- [ ]     A few suggestions to make things prettier.
- [ ]     There are a few nits with `setup_environment`, `setup_dependent_environment`, and `setup_extension_environment` I'm not sure about. I think we probably need more than one method here b/c the env info goes in separate module files.  I think the choice to factor out the module-scope modifications as you have done is a good one, too.
- [ ]     We need to figure out how to get compiler wrapper variables like `OMPI_CC` to behave differently in and out of Spack.  In Spack they should use Spack's compiler wrappers.  Outside of Spack they should call real compilers.

I think that is it.  Thanks!

File: lib/spack/spack/package.py (hunk 5 above)
Comment (tgamblin) [inline lib/spack/spack/package.py]: I like the first for this stuff b/c it's less typing for the package author (who ideally doesn't have to know Python or OO really well).  The second would make sense to me if it was a more complicated framework and they could ever return something other than an `EnvironmentModifications` but here it just seems like extra boilerplate.

File: lib/spack/spack/package.py (hunk 6 above)
Comment (tgamblin) [inline lib/spack/spack/package.py]: @alalazo: good point.  `setup_dependent_environment` makes sense then.  The package author can check whether it's modifying an extension or not by querying the `dependent_spec`, anyway.

File: lib/spack/spack/environment.py (hunk 11 above)
Comment (tgamblin) [inline lib/spack/spack/environment.py]: So maybe keep it and do like this?

class NameValueModifier(object):
    def __init__(self, name, value, **kwargs):
        self.name = name
        self.value = value
        self.format_args = { 'name' : name, 'value' : value, 'path' : value }
        self.format_args.update(kwargs)

class SetEnv(NameValueModifier):
    def execute(self):
        os.environ[self.name] = str(self.value)

File: lib/spack/spack/package.py [hunk 13]
```diff
@@ -1217,6 +1236,15 @@ def rpath_args(self):
...
 
 
+class PythonExtension(Package):
+    def setup_dependent_environment(self, env, dependent_spec):
+        pass
+
+    def setup_environment(self, env):
+        site_packages = glob.glob(join_path(self.spec.prefix.lib, "python*/site-packages"))
+        if site_packages:
+            env.prepend_path('PYTHONPATH', site_packages[0])
```
Comment (alalazo) [inline lib/spack/spack/package.py]: @tgamblin I tried to solve the issue on `PYTHONPATH` with a base class that just implements the two setup functions for python extensions (an example of use is in `py-nose`). If this is fine with you I'll proceed modifying the rest of the packages.

**Comment by citibeth** [comment]
@alalazo Not sure if this is relevant here.  But... not all packages that install Python extensions use Python build tools (see ibmisc, for example).  I chose to do all build/installation in CMake.  It turns out to be easier to learn one build system (CMake) than two and trying to glue them together (CMake + SetupTools).  Even if this was a poor choice on my part... it is my belief that Spack should be able to install whatever crazy stuff is out there.

I previously wanted customization in module files: my CMake build installed my Python extensions in prefix/pylib, with the intention that you add prefix/pylib to your PYTHON_PATH.  When I discovered that there was no easy way to make Spack create module files that did this, I changed my build to put the python extension in a more standard location.

**Comment by alalazo** [comment]
@citibeth I guess your case will be covered now. If `PYTHONPATH` needs to be set both at install time and within the module to the same value, then you can just write something like : 
```python
def setup_environment(self, env):
    ...
    env.prepend_path('PYTHON_PATH', <custom-path-goes-here>)
    ...
```
If you need a different environment to operate within spack you need to override also:
```python
def setup_dependent_environment(self, env, dependent_spec):
    ...
    env.prepend_path('PYTHON_PATH', <path-to-operate-within-spack>)
    ...
```

File: lib/spack/spack/package.py (hunk 13 above)
Comment (tgamblin) [inline lib/spack/spack/package.py]: I think you don't actually need this.

The `setup_dependent_environment` of extendees is called automatically for extensions [here](https://github.com/LLNL/spack/pull/552/files#diff-c8f14b8e4ac7661fb9ca8ecc00bc2ebcR286), so the `site-packages` and `PYTHONPATH` directories will already be set up automatically before `install()`.  This is actually necessary for `setuptools` to do the right thing.

Probably what should happen is that `setup_dependent_environment` should use the `env` object to set `PYTHONPATH` -- then it will be included in the module and in the `install()` environment.  Extensions can obviously implement their own `setup_environment` if needed, but this way they don't have to do so just for the basic `PYTHONPATH`.

The other thing about the code above is that it makes the extension writer have to remember two things instead of one:

  1. add `extends('python'), and
  2. remember to derive from `PythonExtension`

I think only the first should be necessary.  I could see users getting frustrated if they do 1 and not 2.  Am I missing something?

**Comment by citibeth** [comment]
I see that this involves inheriting from a subclass of Package.  That could conflict with PR #543, which also has users inheriting from a subclass of Package.  I'm not going to say which one of the two is "right" or "needs" the subclassing capability more.  But this issue does need to be thought through.

**Comment by tgamblin** [comment]
@alalazo: careful -- `setup_dependent_environment` is called on the *extendee* before the *extension's* install.  It's a way for Python to bundle extension logic for "standard" extensions.

In @citibeth's `pylib` case, I think that would just be a package that `depends_on` but does not `extend` python, because the `pylib` directory is nonstandard and it doesn't make sense to symlink it into the Python install.  In this case she'd just manually override `setup_environment` in her package to set `PYTHONPATH` using the `env` object.  Then it appears in the module, but because it doesn't *extend* *python*, the standard `python2.x/site-packages` won't be created or added to `PYTHONPATH`.  I assume you would only ever want to load a package like this with a module.

If she converts that to `python2.x/site-packages`, then it's standard again and it can be symlinked into the interpreter prefix, and it would make sense to use `extends`.

File: lib/spack/spack/package.py (hunk 13 above)
Comment (tgamblin) [inline lib/spack/spack/package.py]: Forgot to mention that the other thing you miss with the subclass is that `setup_dependent_environment` is able to create the proper X.Y in `pythonX.Y/site-packages`, because it's called on the *extendee* (`python`).  So it knows the python version that the sub-package is extending, and can create the right path.  The version above uses glob and assumes someone has already created that directory, but I've seen `setuptools` do the wrong thing when `pythonX.Y/site-packages` doesn't exist in advance.  This is why it's done the way it is.

This is getting into a lot of python-specific stuff but I think the requirements and conventions are similarly complicated for other languages.  Maybe not quite as combinatorially versioned as python though...

Comment (alalazo) [inline lib/spack/spack/package.py]: @tgamblin I think I may be missing some bits... For `PYTHONPATH` I am trying to reproduce the behavior of develop, but moving the extension specific knowledge out of `modules.py`. The code that in develop takes care of adding `PYTHONPATH` in `modules.py` is:
```python
if self.spec.name != 'python':
    site_packages = glob(join_path(self.spec.prefix.lib, "python*/site-packages"))
    if site_packages:
        add_path('PYTHONPATH', site_packages[0])
```
This logic is triggered only for modules (be it dk or tcl), because as you said, when things are installed it's the `setup_dependent_environment` of python that kicks in:
```python
# Set PYTHONPATH to include site-packages dir for the
# extension and any other python extensions it depends on.
python_paths = []
for d in ext_spec.traverse():
    if d.package.extends(self.spec):
        python_paths.append(os.path.join(d.prefix, self.site_packages_dir))
    os.environ['PYTHONPATH'] = ':'.join(python_paths)
```
The class above should be a way to reproduce the same behavior and transfer the responsibility to the packages that implements python extensions. At install time it's the `setup_dependent_environment` of python that is executed, and the implementation it's the same as the one in develop... so I believe the code here should behave as correctly as in develop. Do you see a fault in the reasoning?

As far as convenience is concerned I agree with you that this is not the optimal solution. I guess anyhow that for most python packages the implementation of `setup_environment` will be the same and given the number of packages already in spack we should definitely find a way of factoring the code in a unique place. The only idea that comes to my mind right now would be patching the default implementation of `setup_environment` and `setup_dependent_environment` on the fly when an `extends` directive is used. This is also kind of nasty, and if you have a better idea I will be glad to implement it :smile:

**Comment by alalazo** [comment]
@citibeth the subclass is meant for "standard" extension, and is an attempt to factor code that otherwise would be repeated over and over. It doesn't give you anything more than two common overrides for `setup_environment` and `setup_dependent_environment`.

@tgamblin @citibeth sorry, I think I wasn't clear enough above, so I'll try to be more schematic:

 1. `setup_dependent_environment(self, env, *dependent_spec*)` will be called from within spack when something that **depends on** or **extends** your package is getting installed
 2. `setup_environment(self, env)` will instead be called when module files are written for your package
 3. in most cases I think you don't need to modify the environment in different ways for 1. and 2. and you don't need to know who is the dependent so the default behavior for `setup_dependent_environment(self, env, *dependent_spec*)` is to call  `setup_environment(self, env)`

**Comment by citibeth** [comment]
> the subclass is meant for "standard" extension

Hmm... your subclass provides useful features when building Python (and other) extensions.  My subclass provides useful features when building packages that use CMake.  What happens when we have a package that uses CMake to build Python extensions (as I do)?  Something must give here.... some kind of mix-in or something is needed.

File: lib/spack/spack/package.py (hunk 13 above)
Comment (alalazo) [inline lib/spack/spack/package.py]: @tgamblin I forgot to say that if you believe `PYTHONPATH` is an issue that needs more work / thinking I can revert the bits of code above into `inspect_path` in `modules.py` and leave the issue for another PR.

File: var/spack/repos/builtin/packages/openmpi/package.py (hunk 9 above)
Comment (alalazo) [inline var/spack/repos/builtin/packages/openmpi/package.py]: @tgamblin I think this is solved by the splitting of `setup_environment` and `setup_dependent_environment`, see `openmpi` or `mpich` packages in this PR for an example of use.

File: lib/spack/spack/package.py (hunk 13 above)
Comment (alalazo) [inline lib/spack/spack/package.py]: @tgamblin Ok, sorry for the noise, but I think I may have it. What if we patch the `setup_environment` and `setup_dependent_environment` of python extensions from:
```
def modify_module(self, module, spec, dependent_spec):
...
```
of the `python` package, if:
- the functions were not overridden explicitly by `dependent_spec`
- `dependent_spec` **extends** python 

After all `modify_module` seems the right place to do those kind of things...

**Comment by citibeth** [comment]
Maybe we can get clever with mix-ins (i.e. multiple inheritance).

https://www.ianlewis.org/en/mixins-and-python

-- Elizabeth

**Comment by alalazo** [comment]
@citibeth I guess not in a package : actually @tgamblin made a very good point above of maintaining code writing simple for packages. I think that in this PR I will either find a way to patch python extension packages or I'll revert to the previous behavior.

Then : it's not much I code in python (so take what I say now with a grain of salt), but I guess that using mixins will require some strict conventions on how method calls are coded. The first thing that comes to my mind is a pervasive use of `super` whenever you override a call to avoid breaking the MRO chain. 

In the example you posted e.g. it makes no sense to add `Mixin1` as it adds nothing to behavior. If `Mixin1` and `Mixin2` behaviors happen to be orthogonal (i.e. they are composable), then the `test` method should have been written like : 
```
def test():
    super().test()
    print <tag>
```
The same is true for every method overridden by the mixin. Order starts to count in your inheritance and in the placement of the `super()` call within the method, if somebody doesn't call `super` it breaks the chain of calls for anything that comes after him in the MRO... I am not sure we should ask to know such things to package maintainers...

**Comment by citibeth** [comment]
I agree, multiple inheritance is distasteful.  Although maybe it would be
OK if we set up a pattern and just tell users to use it.

An alternative would be to always inherit from Package and use monkey
patching to provide mix-in capability.  For example, suppose we have a
command use() that finds all methods defined in a mix-in and monkey-patches
them into the current class instance.  Then I could write:

```
class MyPackage(Package):
    def __init__(self):
        self.use(CMakePackage)
        self.use(DependentEnvPackage)
    ...
```


Or alternatively, maybe we can set things up so we do this at the class
level, rather than instance level (probably possible):


```
class MyPackage(Package):
        use(CMakePackage)
        use(DependentEnvPackage)
    ...
```
-- Elizabeth

File: var/spack/repos/builtin/packages/python/package.py [hunk 14]
```diff
@@ -103,6 +114,31 @@ def setup_dependent_environment(self, module, spec, ext_spec):
...
+                (name, defining_cls) for name, _, defining_cls, _, in inspect.classify_class_attrs(pkg_cls)
+                if
+                name in interesting_overrides and  # The attribute has the right name
+                issubclass(defining_cls, Package) and defining_cls is not Package  # and is an actual override
+            ]
+            if not overrides_found:
+                # If no override were found go on patching
+                pkg_cls.setup_environment = functools.wraps(Package.setup_environment)(_setup_env)
+                pkg_cls.setup_dependent_environment = functools.wraps(Package.setup_dependent_environment)(_setup_denv)
+
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/python/package.py]: @tgamblin This should solve the problem we discussed yesterday regarding not having to inherit from a class when writing extensions. Basically : it's the extendee that will patch an extension where it seems appropriate to do so. Now all the python modules should be writing the same `PYTHONPATH` as in develop without having to modify a line of their code. I tested that manually with `py-nose` and it seems to work fine.

File: var/spack/repos/builtin/packages/python/package.py [hunk 15]
```diff
@@ -103,6 +114,31 @@ def setup_dependent_environment(self, module, spec, ext_spec):
         else:
             module.python = Executable(join_path(spec.prefix.bin, 'python'))
 
+        # The code below patches the any python extension to have good defaults for `setup_dependent_environment` and
+        # `setup_environment` only if the extension didn't override any of these functions explicitly.
+        def _setup_env(self, env):
+            site_packages = glob.glob(join_path(self.spec.prefix.lib, "python*/site-packages"))
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/python/package.py]: @tgamblin I left the logic here exactly the same it was in develop. Please tell me if you want to change it with something more appropriate.

File: lib/spack/spack/modules.py [hunk 16]
```diff
@@ -156,18 +159,36 @@ def write(self):
...
 
-        # If there are no paths, no need for a dotkit.
-        if not self.paths:
+        # Environment modifications guessed by inspecting the installation prefix
+        env = inspect_path(self.spec.prefix)
+
+        # Let the extendee modify their extensions before asking for package-specific modifications
+        for extendee in self.pkg.extendees:
+            extendee_spec = self.spec[extendee]
+            extendee_spec.package.modify_module(self.pkg.module, extendee_spec, self.spec)
```
Comment (alalazo) [inline lib/spack/spack/modules.py]: @tgamblin Just for completeness, the latest approach required to add these three lines to the module file writer

**Comment by alalazo** [comment]
@tgamblin I think I am done with the comments of your review. If you like the approach above I'll just clean the few `TODO` comments I left behind as a reminder and I will be ready to merge.

By the way, thanks for taking the time to review the code : I see that it's a one versus many battle and I imagine it's taking some effort!

File: lib/spack/spack/cmd/uninstall.py [hunk 17]
```diff
@@ -94,11 +94,11 @@ def num_installed_deps(pkg):
         for pkg in pkgs:
             try:
                 pkg.do_uninstall(force=args.force)
-            except PackageStillNeededError, e:
+            except PackageStillNeededError as e:
                 tty.error("Will not uninstall %s" % e.spec.format("$_$@$%@$#", color=True))
-                print
-                print "The following packages depend on it:"
+                print()
```
Comment (tgamblin) [inline lib/spack/spack/cmd/uninstall.py]: @alalazo: this prints `()` in Python 2.6.

**Comment by tgamblin** [comment]
@alalazo: I just pushed a branch with some additions to this that I think makes the interface a bit cleaner and avoids the need to patch anything.  It's in `epfl-scitas-features/env_objects_flying_around`.  See what you think.

Basically the interface looks like this now:

```python
    def setup_environment(self, spack_env, run_env)
    def setup_dependent_environment(self, spack_env, run_env, dependent_spec)
    def setup_dependent_python_module(self, module, dependent_spec)
```

The intent is to make it clear when we're modifying the compile environment and when we're modifying the runtime/module environment.  To make it clear, the two env methods have two environment module objects: `spack_env` and `run_env`.  One modifies the compile environment and the other modifies the run environment, and changes to the run env will go into modules.  It was possible to modify the compile env for a package before by simply setting `env['foo'] = 'bar'` within `install()`, but I think this factoring is nicer because it allows the package author to use the env modifications the same as all the other env settings if they like.

`modify_module` is renamed to `setup_dependent_python_module` to be consistent with the other methods, and to distinguish the python module scope from environment modules in the name.

I also tried to rename `set_env` and `unset_env` to `set` and `unset` in `EnvironmentModifications` (because the `_env` was repetitive most places it was used) and added `set_path` to simplify dealing with paths (user doesn't have to call `concatenate_paths`).

Anyway, see what you think.  I think separating this a little more makes the code more clear and avoids the need to patch.

**Comment by alalazo** [comment]
@tgamblin I really like passing two different environments, in particular as it permits to avoid monkey patching the code (a behavior which would not have been obvious from user perspective). I also think that all the modifications to `EnvironmentModifications` effectively make the API more usable.

There are a few thoughts / remarks that I would like to share : 

 1. is it still necessary to have `setup_environment` at all? Before the idea was that the run-time environment was responsibility of `setup_environment`, while the spack environment was on `setup_dependent_environment`. Now that two `env` instances are passed to `setup_dependent_environment` is the other function still needed?

2. I would rename `setup_dependent_python_module` to `setup_dependent_module`, just to avoid people confusing the python module you are referring to with the python package provided by spack

3. there are a few missing things in `openmpi` and `mpich` (basically setting `run_env`) + `mpich` call seem to have a wrong signature

If you agree with point 1. in particular I can pull your modifications here and try to simplify the PR a bit further.

**Comment by alalazo** [comment]
@tgamblin It seems that also edits to older messages are not notified to me. I just saw accidentaly your PS above (and that's kind of a shame, as I edit constantly my posts). If it may work for you, I've issued an invitation to our organization so to let you being able to push on PR branches.

**Comment by alalazo** [comment]
@tgamblin On a second thought having both calls may be indeed useful. Here's the use case that comes to my mind : `hdf5+mpi`. I think its tcl module file should have `depends_on <concrete-mpi>` and it's the actual mpi that should add that to the `run_env` in `setup_dependent_environment`. If you agree I'll try to implement this case here.

**Comment by tgamblin** [comment]
@alalazo:

  1. yes I think we both agree now that we need both.  One allow you to set up *your* package's environment, the other allows you to set up something in the environment of *dependencies*.  The Python case actually makes use of this -- extensions' modules need to add *their own* PYTHONPATH to the environment, but it depends on a) the version of Python (the extendee) and b) the prefix of the extension.

2. I'm not opposed to `setup_dependent_module`.  I do worry that people will be confused by this, though, as they'll thing `module` refers to a TCL/Lmod module.  What if we called it `setup_dependent_package`?  `Package` at least is not quite as overloaded and you're really setting variables to be used by the dependent package.

3. I was going to comment on the OMPI and MPICH stuff.  I took out the `run_env` settings there because the compilers are patched by `filter_compilers`.  You don't actually need to have a module set `OMPI_CC` or `MPICC_CC`.  Do you think the env vars should be set anyway?

> I think its tcl module file should have depends_on <concrete-mpi> and it's the actual mpi that should add that to the run_env in setup_dependent_environment.

I think this is useful, although the ORNL folks (@robertdfrench, others) mentioned that they have some more generic prereq logic that they would like to see added.  Do you think this capability could be added later if this PR is mostly ready to go?  I know @jgalarowicz would appreciate the capabilities that are already in here.

Aslo, did you see #498?  Is this a decent naming scheme?  It would be nice to support @glennpj's use case here, and I think his module names are better than the current builtin ones.

**Comment by alalazo** [comment]
>I think this is useful, although the ORNL folks (@robertdfrench, others) mentioned that they have some more generic prereq logic that they would like to see added. Do you think this capability could be added later if this PR is mostly ready to go? 

Yes. Next on my list for module files I have : 

- support for LMOD (i.e. update #107 so that it will be ready to be merged)
- support for custom naming scheme (I'll take a close look at #498 tomorrow, maybe iterate with @glennpj on this)
- support for site-customization (i.e. add an hook to read variables from a configuration file)

and having this PR merged would surely help to proceed.

> What if we called it setup_dependent_package

Ok.

**Comment by tgamblin** [comment]
@alalazo: awesome!  So is this one ready to be merged?

**Comment by alalazo** [comment]
@tgamblin Yes, the only detail I am wondering is if in `modules.py` we should iterate only on extendees (I originally meant it for python and the likes) or on all the dependencies. Maybe the latter is more appropriate right now. Besides that I think we can merge.

**Comment by tgamblin** [comment]
@alalazo: I think the Python package should probably iterate only on extendees, but I don't think we should add another callback function for that.  Packagers can call `dependent_spec.package.is_extension` to check for this if they only want to affect extendees.

**Comment by alalazo** [comment]
@tgamblin Ok then. Merge! :smile:

**Comment by tgamblin** [comment]
Thanks for all the work on modules!

**Comment by tgamblin** [comment]
@robertdfrench @jgalarowicz: you should be able to add some variables to the generated module files now.

**Comment by tgamblin** [comment]
TODO: docs.

**Comment by tgamblin** [comment]
Oops --  forgot to merge *my* changes here.  Sigh.  I'll put those in.

**Comment by alalazo** [comment]
@tgamblin Eheh, I thought you decided to merge this and then your changes from llnl branch...

**Comment by tgamblin** [comment]
Fixed.  :relaxed:
